File: core/src/com/unibo/s3/testbed/testbed_modules/EntitiesSystem.java
package com.unibo.s3.testbed.testbed_modules;

import com.badlogic.gdx.math.Vector;
import com.unibo.s3.main_system.characters.steer.MovableEntity;

import java.util.List;

public interface EntitiesSystem<T extends Vector<T>> {

    MovableEntity<T> spawnEntityAt(T position);

    void spawnEntity(MovableEntity<T> newEntity);

    List<MovableEntity<T>> getEntities();

    Iterable<MovableEntity<T>> getNeighborsOf(MovableEntity<T> entity, float searchRadius);
}

File: core/src/com/unibo/s3/main_system/characters/EntitiesSystem.scala
package com.unibo.s3.main_system.characters

import com.badlogic.gdx.math.Vector2
import com.unibo.s3.main_system.communication.CharacterActors

/**
  * This trait is used to generate a new character
  * @author Daniele Rosetti
  */
trait EntitiesSystem {
  type CharacterActors = CharacterActors.Value

  /**
    * Generate a new BaseCharacter
    * @param character Type of character to generate
    * @param position Place where new character is spawn
    * @param Id A number to identify the BaseCharacter
    * @return New BaseCharacter
    */
  def spawnEntityAt(character: CharacterActors, position: Vector2, Id: Int): BaseCharacter

  /**
    * this method return all character
    * @return All BaseCharacter
    */
  def getEntities: Iterable[BaseCharacter]
}
